integration/storagetest/logtests.go

```diff
@@ -453,3 +451,85 @@ func createFakeLeaf(ctx context.Context, s storage.LogStorage, tree *trillian.Tr
 
 	return q[0].Leaf
 }
+
+func (*logTests) TestDequeueLeaves(ctx context.Context, t *testing.T, s storage.LogStorage, as storage.AdminStorage) {
+	const leavesToInsert = 5
+	tree := mustCreateTree(ctx, t, as, storageto.LogTree)
+	mustSignAndStoreLogRoot(ctx, t, s, tree, 0)
+
+	leaves := createTestLeaves(leavesToInsert, 20)
+	if _, err := s.QueueLeaves(ctx, tree, leaves, fakeDequeueCutoffTime); err != nil {
+		t.Fatalf("Failed to queue leaves: %v", err)
+	}
+
+	// Now try to dequeue them
+	// Some dequeue implementations probabalistically dequeue and require retrying until timeout.
+	leaves2 := dequeueLeaves(ctx, t, s, tree, fakeDequeueCutoffTime, 5, 0)
+	if len(leaves2) != leavesToInsert {
+		t.Fatalf("Dequeued %d leaves but expected to get %d", len(leaves2), leavesToInsert)
+	}
+
+	// If we dequeue again then we should now get nothing
+	if err := s.ReadWriteTransaction(ctx, tree,
+		func(ctx context.Context, tx3 storage.LogTreeTX) error {
+			leaves3, err := tx3.DequeueLeaves(ctx, 99, fakeDequeueCutoffTime)
+			if err != nil {
+				t.Fatalf("Failed to dequeue leaves (second time): %v", err)
+			}
+			if len(leaves3) != 0 {
+				t.Fatalf("Dequeued %d leaves but expected to get %d", len(leaves3), leavesToInsert)
+			}
+			return nil
+		}); err != nil {
+		t.Fatalf("Could not dequeue the expected number of leaves: %v", err)
+	}
+}
+
+// dequeueLeaves starts a transaction and uses it to dequeue leaves.
+func dequeueLeaves(ctx context.Context, t *testing.T, ls storage.LogStorage, tree *trillian.Tree, ts time.Time, limit int, startIndex int64) []*trillian.LogLeaf {
+	// We'll retry a few times if we get nothing back since we're now dependent
+	// on the underlying queue delivering unsequenced entries.
+	var ret []*trillian.LogLeaf
+	cctx, cancel := context.WithTimeout(ctx, 5*time.Second) // Retry until timeout
+	defer cancel()
+	err := ls.ReadWriteTransaction(cctx, tree, func(ctx context.Context, tx storage.LogTreeTX) error {
+		ret = make([]*trillian.LogLeaf, 0, limit)
+		i := 0
+		start := time.Now()
+		for len(ret) < limit {
+			i++
+			got, err := tx.DequeueLeaves(ctx, limit, ts)
+			if err != nil {
+				return err
+			}
+			ret = append(ret, got...)
+		}
+		t.Logf("DequeueLeaves took %d tries and %v to dequeue %d leaves", i, time.Since(start), len(ret))
+		ensureAllLeavesDistinct(ret, t)
+		iTimestamp := ptypes.TimestampNow()
+		for i, l := range ret {
+			l.IntegrateTimestamp = iTimestamp
+			l.LeafIndex = int64(i) + startIndex
+		}
+		if err := tx.UpdateSequencedLeaves(ctx, ret); err != nil {
```

Comment:
Given that this does sequencing in addition to dequeueing, I'd rename the function to something like `dequeueAndSequence`?

Reply:
done